### PR TITLE
[#68223] add numeric representation helper

### PR DESCRIPTION
**Pull request by Kharonus**
# Ticket
[#68223](https://community.openproject.org/wp/68223)

# What are you trying to accomplish?
- get a functional component, that can handle very big or very small numbers and render them accordingly
- the function takes options about precision, size and overall length of the string representation and falls back to scientific notation IF any of the constraints is violated

# What approach did you choose and why?
- add helper for formatting numbers
- add tests to express expectations

# Hint
- check test cases for the expected functionality
- raise a PR comment, if the expectation wouldn't match with your expectations

### CODE REVIEW

File: app/helpers/number_format_helper.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+  def integer_part_size(number) = number.round.to_s.size
+
+  def format_scientific_notation(number, precision)
+    exponent = number.to_d.exponent
+    rounding_position = exponent - precision
+    number.round(-1 * rounding_position).to_d.to_s("E")
+  end
+
+  def init_formatting_options(opts)
```
Comment (NobodysNightmare) [inline app/helpers/number_format_helper.rb:67]: :yellow_circle: Why do we use a specific helper function here that initializes `opts` instead of using keyword-arguments with default values in the first place?

File: app/helpers/number_format_helper.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    {
+      delimiter: I18n.t("number.format.delimiter"),
+      separator: I18n.t("number.format.separator"),
```
Comment (NobodysNightmare) [inline app/helpers/number_format_helper.rb]: :yellow_circle: These two values don't quite fit well IMO. They are hardcoded I18n translations. Why don't we use:

* a local variable `delimiter = I18n.t("number.format.delimiter")` **or**
* a shared private method `def delimiter`

File: app/helpers/number_format_helper.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  end
+
+  private
+
+  def integer_part_size(number) = number.round.to_s.size
+
+  def format_scientific_notation(number, precision)
+    exponent = number.to_d.exponent
+    rounding_position = exponent - precision
+    number.round(-1 * rounding_position).to_d.to_s("E")
```
Comment (NobodysNightmare) [inline app/helpers/number_format_helper.rb]: :green_circle: I am confused enough by this code, that I might wish for a short comment explaining what's going on here. Most things are clear to me: We format a `BigDecimal` using `to_s("E")` which enforces scientific notation. Before that we want to use rounding, to ensure that the precision is what we need.

However, I don't understand how the rounding works:

* Why multiply by `-1`?
* How do exponent and precision work together here?

File: spec/helpers/number_format_helper_spec.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require "spec_helper"
+
+RSpec.describe NumberFormatHelper do
+  describe "#number_with_limit" do
+    [
+      { input: 10, opts: {}, result: "10" },
+      { input: 10.0, opts: {}, result: "10" },
```
Comment (NobodysNightmare) [inline spec/helpers/number_format_helper_spec.rb:38]: :yellow_circle: There are no test cases for the default options that force scientific notation or slightly don't require scientific notation.

I'd have expected test cases like "the biggest default number without scientific notation", e.g.

```ruby
{ input: 10_000_000, opts: {}, result: "10,000,000" 
```

edit: After reading through all the test cases, I see a general lack of boundary tests from the non-scientific end. What I'd expect is a pair of test cases for each of the limits we defined:

* the "biggest" non-scientific
* the "smallest" scientific
* Each of those pairs for
    * length_limit
    * digits
* if applicable for the limit, I'd probably do this for a large number (10000000) and for a small number (0.000001)

File: spec/helpers/number_format_helper_spec.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      { input: 17203230.09932, opts: { length_limit: 8, digits: 8, precision: 3 }, result: "0.172e8" },
```
Comment (NobodysNightmare) [inline spec/helpers/number_format_helper_spec.rb]: :green_circle:  Ruby allows inserting `_` into numbers to make them more readable. This doesn't change the represented number in any way, but can make code with large numbers more readable. E.g. in this line

```suggestion
      { input: 17_203_230.09932, opts: { length_limit: 8, digits: 8, precision: 3 }, result: "0.172e8" },
```

I could easily see that the number is in the tens of millions.

File: spec/helpers/number_format_helper_spec.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      { input: 17203230.09932, opts: { length_limit: 20, digits: 6, precision: 4 }, result: "0.172e8" },
+      { input: 17203230.09932, opts: { length_limit: 20, digits: 6, precision: 5 }, result: "0.17203e8" },
+      { input: 17203230.09932, opts: { length_limit: 20, digits: 10, precision: 3 }, result: "17203230.099" },
```
Comment (NobodysNightmare) [inline spec/helpers/number_format_helper_spec.rb]: :yellow_circle: I am surprised by the expected result. Why do we not expect a thousands separator to be part of the string? Do we really want to format large numbers without that?

```suggestion
      { input: 17203230.09932, opts: { length_limit: 20, digits: 10, precision: 3 }, result: "17,203,230.099" },
```

File: spec/helpers/number_format_helper_spec.rb (hunk 5 above)
Comment (NobodysNightmare) [inline spec/helpers/number_format_helper_spec.rb]: :yellow_circle: Why are `length_limit` and `digits` different in this test compared to the two following test cases?

The only significant difference seems to be the `precision`. The input number and the expected format stay the same. Since none of these specs have a written expectation (e.g. "when the number is longer than the allowed number of digits"), it's hard to follow why a certain line has certain expectations. If input values then change without any meaningful influence on the result, it makes things more confusing.

In case this test was testing more than one thing: Split the test in two and let's make sure there's one test per expectation.

E.g. what I see between this line and the two next lines is the influence of precision:

* precision 3 and 5 give 3 and 5
* precision 4 is an interesting edge case, because we don't want the number to end in "0", so it gives precision 3 as well

File: spec/helpers/number_format_helper_spec.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      { input: 8.402e-13, opts: { length_limit: 9, digits: 6, precision: 2 }, result: "0.84e-12" },
```
Comment (NobodysNightmare) [inline spec/helpers/number_format_helper_spec.rb]: :yellow_circle: This spec highlights something interesting and I nearly didn't notice it.

I'd usually expect a significant number in front of the comma. You apparently expect this as well, that's why you wrote `8.402e-13` and not `0.8402e-12`. Though we expect the latter in the final formatting.

I guess the form chosen might make sense. For example the large number `0.172e8` is 8 digits long and `e8` shows that to me directly. So maybe it's even better.

Though in that case, I'd suggest to at least write specs where we input the number in scientific notation with the same exponent as what we expect later on.

File: app/helpers/number_format_helper.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      precision:,
+      length_limit:
+    }
+
+    string_number = number_with_precision(number, precision:, strip_insignificant_zeros: true)
+    length = string_number.delete("#{delimiter}#{separator}").length
+
+    scientific_notation_needed = length > length_limit ||
+                                 integer_part_size(number) > digits ||
+                                 (string_number == "0" && number != 0)
```
Comment (NobodysNightmare) [inline app/helpers/number_format_helper.rb:40]: This looks like a very specific edge case, I guess for very small numbers that get rendered as "0".

Do we have a proper test for this case?  I can't identify one.

What I'd expect to be tested:

* "0.0" (a floating point zero) can still be rendered as "0" very normally
* a number for which this rule was added and that's not caught by other rules is rendered scientifically

**Review by NobodysNightmare — CHANGES_REQUESTED** [review]

File: app/helpers/number_format_helper.rb (hunk 2 above)
Comment (Kharonus) [reply app/helpers/number_format_helper.rb]: Is there a big difference then having them in the opts hash? ^^ I have no big preference here, what is your preferred way?

File: app/helpers/number_format_helper.rb (hunk 3 above)
Comment (Kharonus) [reply app/helpers/number_format_helper.rb]: will explain it with a comment

File: spec/helpers/number_format_helper_spec.rb (hunk 5 above)
Comment (Kharonus) [reply spec/helpers/number_format_helper_spec.rb]: good hint

File: app/helpers/number_format_helper.rb (hunk 2 above)
Comment (NobodysNightmare) [reply app/helpers/number_format_helper.rb]: I am somewhere between using it inline and if that doesn't work for readabiltiy purposes, I'd probably move it out into a method `def delimiter` and `def separator`.

Comment (NobodysNightmare) [reply app/helpers/number_format_helper.rb]: The difference to having it in the opts hash IMO is semantics. This method here is used to parse the `opts` hash and the returned thing has nothing to do with the opts hash.

It's just other useful "stuff".

File: spec/helpers/number_format_helper_spec.rb (hunk 6 above)
Comment (Kharonus) [reply spec/helpers/number_format_helper_spec.rb]: you probably mean a delimiter :D And yes, this is the right question. A question we need to ask product.

File: spec/helpers/number_format_helper_spec.rb (hunk 5 above)
Comment (Kharonus) [reply spec/helpers/number_format_helper_spec.rb]: This comes from requirements. The `length_limit` constraints the length of the entire number, digits before and after separator. the `digits` are only referring to the digits before the comma.

File: spec/helpers/number_format_helper_spec.rb (hunk 7 above)
Comment (Kharonus) [reply spec/helpers/number_format_helper_spec.rb]: I found no way of having ruby formatting `BigDecimal` in scientific notation with the significant digit before the comma. :/

File: spec/helpers/number_format_helper_spec.rb (hunk 5 above)
Comment (NobodysNightmare) [reply spec/helpers/number_format_helper_spec.rb]: Yeah, I understand that `length_limit` and `digits` do that. What I mean is these three test cases:

```ruby
{ input: 17203230.09932, opts: { length_limit: 8, digits: 8, precision: 3 }, result: "0.172e8" },
{ input: 17203230.09932, opts: { length_limit: 20, digits: 6, precision: 4 }, result: "0.172e8" },
{ input: 17203230.09932, opts: { length_limit: 20, digits: 6, precision: 5 }, result: "0.17203e8" }
```

* they have the same input
* test 2&3 have the same length limit, test 1 is different
* test 2&3 have the same digits, test 1 is different
* all tests have different precision
* the difference in `result` is due to `precision` for all three test cases, length_limit and `digits` have no influence

And that's what I find confusing here. Because reading the cases for the first time, it's not clear what the difference between the specs is.

So my specific suggestion is

```ruby
{ input: 17203230.09932, opts: { length_limit: 20, digits: 6, precision: 3 }, result: "0.172e8" },
{ input: 17203230.09932, opts: { length_limit: 20, digits: 6, precision: 4 }, result: "0.172e8" },
{ input: 17203230.09932, opts: { length_limit: 20, digits: 6, precision: 5 }, result: "0.17203e8" }
```

and then having other test cases that test differences in behaviour for `length_limit` and `digits`.

File: app/helpers/number_format_helper.rb (hunk 8 above)
Comment (Kharonus) [reply app/helpers/number_format_helper.rb:40]: actually this case is very common. It happens everytime, the precision is smaller, then the exponent (abs) for a small number.

`number_with_precision` will always render 0, if within the precision only zeros are found.

Comment (Kharonus) [reply app/helpers/number_format_helper.rb:40]: The test cases here show the difference:

```ruby
      { input: 8.402e-13, opts: { length_limit: 20, digits: 6, precision: 20 }, result: "0.0000000000008402" },
      { input: 8.402e-13, opts: { length_limit: 9, digits: 6, precision: 4 }, result: "0.8402e-12" }
```

I could rewrite the opts of the second to `length_limit: 20`, to emphasize it

Comment (NobodysNightmare) [reply app/helpers/number_format_helper.rb:40]: > It happens everytime, the precision is smaller, then the exponent (abs)

Doh, of course! I forgot that the precision influences the number rendering for the non-scientific case.

> The test cases here show the difference [...] I could rewrite the opts of the second to length_limit: 20, to emphasize it

YES PLEASE! :D This is what I meant in one or two of my other comments, but I'll write it here again: Right now the second test is less useful than it could be. The scientific notation is **either** enabled, because you reduced the precision **or** it's enabled because you reduced the length limit. Both would lead to the same result, but I as a reader have no clue why it happened and anyone that wants to ensure the implementation is working in all the cases also can't rely on the test, because an error in one of the cases would be hidden by the other case working correctly.

Comment (Kharonus) [reply app/helpers/number_format_helper.rb:40]: done

File: app/helpers/number_format_helper.rb (hunk 3 above)
Comment (Kharonus) [reply app/helpers/number_format_helper.rb]: I found a better formatting approach, to satisfy also the want of displaying the first non-zero digit before the comma. And also use all the other nice formatting needs.

File: app/helpers/number_format_helper.rb (hunk 2 above)
Comment (Kharonus) [reply app/helpers/number_format_helper.rb]: extracted to private methods

File: app/helpers/number_format_helper.rb (hunk 1 above)
Comment (Kharonus) [reply app/helpers/number_format_helper.rb:67]: Easy answer: extract complexity out of the single main method. 🤷🏽‍♀️

File: spec/helpers/number_format_helper_spec.rb (hunk 4 above)
Comment (Kharonus) [reply spec/helpers/number_format_helper_spec.rb:38]: added more default cases

File: spec/helpers/number_format_helper_spec.rb (hunk 6 above)
Comment (Kharonus) [reply spec/helpers/number_format_helper_spec.rb]: -> won't introduce it (yet) until someone tells me WHEN to use a delimiter (only in non-scientific notation or both or never)

File: spec/helpers/number_format_helper_spec.rb (hunk 5 above)
Comment (Kharonus) [reply spec/helpers/number_format_helper_spec.rb]: I improved and added more test cases.

File: spec/helpers/number_format_helper_spec.rb (hunk 7 above)
Comment (Kharonus) [reply spec/helpers/number_format_helper_spec.rb]: I now found a way ... using string format and then some normalization

File: app/helpers/number_format_helper.rb [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+  def integer_part_size(number) = number.round.to_s.size
+
+  def format_scientific_notation(number, precision)
+    str = "%0.*e" % [precision, number]
+
+    mantissa, exponent = str.split("e")
+
+    # normalize mantissa
+    mantissa = mantissa.sub(/(\.\d*?)0+$/, '\1')
```
Comment (NobodysNightmare) [inline app/helpers/number_format_helper.rb]: It's a bit hard to understand what the intention of this regular expression is. It might help to describe what exactly we are normalizing here.

Specifically I am wondering about this piece of the regex: `\d*?` The `?` seems superfluous, because `*` already matches zero repetitions of `\d`.

File: app/helpers/number_format_helper.rb [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    mantissa = "#{mantissa}0" if mantissa.end_with?(".")
+
+    # normalize exponent
+    exp_int = exponent.to_i
```
Comment (NobodysNightmare) [inline app/helpers/number_format_helper.rb]: I'd hope us to expect the exponent to always be parseable, so I'd suggest we do `Integer(exponent)`. Still I am wondering which case we are exactly guarding against here. Could the exponent be rendered differently sometimes?

If we don't expect any changes, we could also use `exponent` in the formatting string below unchanged.

File: spec/helpers/number_format_helper_spec.rb [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+      { input: 10, opts: {}, result: "10" },
+      { input: 10.0, opts: {}, result: "10" },
+      # Defaults are converting very big and very small numbers to scientific notation
+      { input: 1_000_000, opts: {}, result: "1000000" },
+      { input: 10_000_000, opts: {}, result: "1.0e7" },
+      { input: 0.0001, opts: {}, result: "0.0001" },
+      { input: 0.00001, opts: {}, result: "1.0e-5" },
+      { input: 100_000.1, opts: {}, result: "100000.1" },
+      { input: 100_000.0001, opts: {}, result: "1.0e5" },
+      { input: 100_000.00001, opts: {}, result: "100000" },
```
Comment (NobodysNightmare) [inline spec/helpers/number_format_helper_spec.rb:46]: Thanks for the three examples above, it took me a moment to understand this edge case properly. It's definitely a weird behaviour, but it's understandable.

I was surprised that the middle number is scientific and the surrounding ones are not, but I guess that's because of the precision threshold being reached.

Good to have this captured in an example.

File: app/helpers/number_format_helper.rb (hunk 1 above)
Comment (NobodysNightmare) [reply app/helpers/number_format_helper.rb:67]: Not easy to me, because I'd probably find it overall easier to read as arguments. Just so that we do not misunderstand, I'd change

```ruby
def number_with_limit(number, opts = {})
  init_formatting_options(opts) => { digits:, precision:, length_limit: }
```

to

```ruby
def number_with_limit(number, digits: 7, precision: 4, length_limit: 9)
```

IMO that doesn't increase the "complexity" of the main method, but it increases its readability, because the method signature already tells me:

* which named parameters exist (also relevant for callers that use a proper IDE like RubyMine)
* what the default values are

Consider this comment a :green_circle:, if you really prefer the separate method, I wont stop you from using it, but my preference in readability is clearly to move this to the method signature.

File: app/helpers/number_format_helper.rb (hunk 9 above)
Comment (NobodysNightmare) [reply app/helpers/number_format_helper.rb]: So I was also surprised about the next line:

```
mantissa = "#{mantissa}0" if mantissa.end_with?(".")
```

As far as I can tell, the format string above could produce a result like `1.0000e5`, which would look weird and you want to ensure that we always have no trailing zeros, but at the same time always a comma (so maybe one trailing zero for `1.0`.

If that's the goal, I think the second line is not necessary, if we change the sub to:

```suggestion
    mantissa = mantissa.sub(/(\.\d+)0+$/, '\1')
```

This would mean we only match and replace, if there is at least a number **and** a zero after the comma, so we could never replace when there is only the zero after the comma.

If this leads to a too greedy first group (e.g. for `1.00000000`, capturing the regex like `1(.0000000)0`), we could increase the regex complexity slightly, but then have it all in a single safe line, which I'd probably still prefer to the current two-line approach:

```suggestion
    mantissa = mantissa.sub(/(\.\d*[1-9])0+$/, '\1')
```

Here we explicitly state the idea that the last number before the removed zeroes must be a number that's not a zero, but any number of digits can come before that.

File: app/helpers/number_format_helper.rb (hunk 10 above)
Comment (NobodysNightmare) [reply app/helpers/number_format_helper.rb]: 17 hours later, I still don't know if this has any purpose. My current guess is that this is a no-op and we could also entirely skip this step.

**Review by NobodysNightmare — APPROVED** [review]
I have some remarks left on the implementation details (one code style, two regarding the normalization). The specs improved greatly though, I am really happy with those now.

I don't think that any of my comments should be a total blocker, but I'd definitely double check the regex :D

File: app/helpers/number_format_helper.rb (hunk 10 above)
Comment (Kharonus) [reply app/helpers/number_format_helper.rb]: using the `stringf` formatting gives us exponents like `-13`, `-08`, `+01`, and `+12`. Only the first one matches our expectations. So, the trick is to parse that string into an integer, to get rid of a leading `+` and any leading zeros.

Comment (Kharonus) [reply app/helpers/number_format_helper.rb]: What is the implementation difference between `Integer(exponent)` and `exponent.to_i`?

File: app/helpers/number_format_helper.rb (hunk 9 above)
Comment (Kharonus) [reply app/helpers/number_format_helper.rb]: `*?` is not superfluous, but a different matcher then `*` and `?`. It means, it expands lazy.

here is the difference between `/(\.\d+)0+$/` and `/(\.\d*?)0+$/`. Your suggestions grabs the zeros in a row greedy, as you already noticed above. This is not what we want to replace.

The second suggestion won't work for numbers that have only zeros after the comma, as those are not matches at all in the capturing group.

File: app/helpers/number_format_helper.rb (hunk 1 above)
Comment (Kharonus) [reply app/helpers/number_format_helper.rb:67]: Okay, I see the point. I want to throw one more thing into the pot: I think in the next step we should make those defaults configurable with settings. But I can't load Settings in method parameter defaults, can I?

File: app/helpers/number_format_helper.rb (hunk 9 above)
Comment (Kharonus) [reply app/helpers/number_format_helper.rb]: Changed to `/(\.\d+?)0+$/` .

File: app/helpers/number_format_helper.rb (hunk 10 above)
Comment (NobodysNightmare) [reply app/helpers/number_format_helper.rb]: One has stronger expectations/guarantees than the other:

```ruby
irb(main):001> "banana".to_i
=> 0
irb(main):002> Integer("banana")
<internal:kernel>:288:in 'Kernel#Integer': invalid value for Integer(): "banana" (ArgumentError)
```

File: app/helpers/number_format_helper.rb (hunk 1 above)
Comment (NobodysNightmare) [reply app/helpers/number_format_helper.rb:67]: Not sure if you like this or hate this, but you can call methods in those.

So this would be possible:

```ruby
def number_with_limit(number, digits: Settings.number_of_digits, precision: Settings.precision, length_limit: Settings.length_limit)
```

That said, depending on the complexity of this, I would see the point of extracting the config loading into a method, but it's hard to judge today's code using tomorrow's requirements.
